blockprotocol/blockprotocol
# Improve usability of `example-graph.json`

## 🌟 What is the purpose of this PR?

We previously had conflicting approaches for how blocks could supply example data, these were
- the `examples` field in the `block-metadata.json` (or the associated field in `blockprotocol` in the `package.json`)
- the `variants.examples` field  in the`block-metadata.json` (or the associated field in `blockprotocol` in the `package.json`)
- the `variants.properties` field in the `block-metadata.json` (or the associated field in `blockprotocol` in the `package.json`)
- the `example-graph.json` file which could be shipped alongside the block

This PR simplifies some of the options by mostly standardizing on the `example-graph` in most cases. In doing so, it improves the ergonomics of defining an `example-graph` by accepting a script that generates the JSON, rather than just a plain-text file. In doing so, this allows users to benefit from the codegen within blocks, and type checking. 

Trying to centralize on `example-graph` means
- It is possible to create sample block entities which are link entities (previously it'd be impossible to define the `linkData` or the accompanying links)
- It is more extensible, we could allow blocks to provide many entities (e.g. variants) and instead find a way of making the `blockEntityRecordId` configurable or provided in a map
- We can get rid of the `examples` keyword, which simplifies the API surface and reduces the number of decisions users have to make and understand
- If a user wants to start adding more entities to the example graph they don’t have to move from one representation to another, they just extend their existing one
- It allows us to have one place to update if we did want to start supporting multiple example sets (as opposed to variants)

This PR additionally modifies the site and hub to appropriately load the example data shipped with blocks, fixing numerous bugs that had slipped in recently.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1203358502199087/1204080201525383/f) _(internal)_

## 🚫 Blocked by

See unresolved comments, I have opened a few to do with open questions around error handling and validation

## 🔍 What does this change?

- Adjusts the webpack config to produce a CommonJS example-graph script (otherwise `block-scripts` struggles)
- Adjusts the `example-graph` `block-scripts` script to allow for an example-graph script, cleaning up after it
- Removes the `examples` field from block metadata
- Adjusts the block templates to use `example-graph.ts` files instead of `examples`
- Updates the block sandboxing and surrounding code on the Hub to correctly parse the data provided by blocks

## 📜 Does this require a change to the docs?

- Some of the accompanying docs and spec have been updated as part of this

## ⚠️ Known issues

- I am unsure of how we currently validate the files pushed by users, I believe if the metadata is malformed the hub throws a top-level error
- The distinction between `variants` and the other ways of defining `examples` is unclear, we should look to unify them
- It is still impossible to provide link entities under `variants`
- In general the `properties` viewer/editor on the hub isn't very intuitive and we're planning a rework soon. It doesn't update to reflect changes the block makes for example.
- ~~The Snackbar crashes the app~~ (fixed)

## 🐾 Next steps

- TBD

## 🛡 What tests cover this?

- Unsure besides static analysis

## ❓ How to test this?

1. `yarn dev:db`
2. `yarn dev:seed-db`
3. `yarn dev`
4. Check the existing blocks on the hub
5. `cd libs/block-template-react`
6. Run the publishing flow locally
7. Check the block template renders correctly and it takes it from the `example-graph`

## 📹 Demo

<img width="1170" alt="image" src="https://user-images.githubusercontent.com/25749103/228012179-a6e9cb69-0ec7-4aa7-860d-3280bebbe5f9.png">
<img width="1263" alt="image" src="https://user-images.githubusercontent.com/25749103/228012296-fb11025b-af40-4b6f-90b1-d00658a245bd.png">

https://user-images.githubusercontent.com/25749103/228214599-ef65577a-5fb3-4933-ae67-4b93f2a7f8e7.mov
